s3/src/main/scala/akka/stream/alpakka/s3/auth/CanonicalRequest.scala

```diff
@@ -29,12 +36,27 @@ private[alpakka] object CanonicalRequest {
       signedHeadersString(req.headers),
       hashedBody
     )
-  }
 
   def canonicalQueryString(query: Query): String =
-    query.sortBy(_._1).map { case (a, b) => s"${uriEncode(a)}=${uriEncode(b)}" }.mkString("&")
+    query.sortBy(_._1).map { case (a, b) => encodeQueryParams(a, b) }.mkString("&")
 
-  private def uriEncode(str: String) = URLEncoder.encode(str, "utf-8")
+  private def encodeQueryParams(name: String, value: String): String =
+    if (name.contains(' ')) {
+      s"${uriEncode(name.takeWhile(_ != ' ')).replace("%7E", "~")}="
+    } else {
+      s"${uriEncode(name).replace("%7E", "~")}=${uriEncode(value).replace("%7E", "~")}"
+    }
+
+  private def uriEncode(str: String) = if (isAlreadyURLEncoded(str)) str else URLEncoder.encode(str, "utf-8")
+
+  // TODO: this could be removed when the Uri class in akka-http will accept utf-8
```

Comment:
https://github.com/akka/akka-http/issues/86 - this is  a similar issue, but for Cyrillic URIs